Hw02 unpack string

zamatay opened the pull request.

Commit:
home work 02

Commit:
home work 02

Commit:
home work 02

Commit:
home work 02

hungry-honeybadger reviewed (commented):
Добрый день. Для начала проверки работы ее нужно сдать через портал.

hungry-honeybadger reviewed (changes requested):
Отличное начало. Круто что не используется встроенный метод IsDigit, у него есть свои проблемы при работе с цифрами. Код отлично отрабатывает с символами ASCII, но он некорректно работает с многосимвольной кодировкой. Это происходит из-за того, что оператор [] со строкой адресует не символ, а байт.

Commit:
home work 02

Commit:
home work 02

zamatay committed:
Merge branch 'master' into hw02_unpack_string

Commit:
add unicode sring to test

Commit:
new solutions

Commit:
new solutions

hungry-honeybadger commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,78 @@
+	return true
+}
+
+func addItem(count int, d *Data) {
```
Рекомендую переделать в метод структуры

hungry-honeybadger commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,78 @@
+	clearResult(d)
+}
+
+func clearResult(d *Data) {
```
Рекомендую переделать в метод структуры

hungry-honeybadger reviewed (changes requested):
Стало намного лучше, но код все еще некорректно обрабатывает некоторые тест кейсы. Например вот этот {input: `\\3qwe\\`, expected: `\\\qwe\`},

Commit:
repair bug!

hungry-honeybadger reviewed (changes requested):
Прошлый баг исчез, но остался еще 1. Тест кейс для демонстрации бага raw строка `asd\` .

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	new file:   go.mod
	modified:   hw02_unpack_string/unpack.go
	modified:   hw02_unpack_string/unpack_test.go
	new file:   main.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go
	modified:   hw02_unpack_string/unpack_test.go

Commit:
del main

Commit:
del main

Commit:
del main

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go

Commit:
Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch hw02_unpack_string
Your branch is up to date with 'origin/hw02_unpack_string'.

Changes to be committed:
	modified:   hw02_unpack_string/unpack.go

Commit:
then end

Commit:
del main

Commit:
del main

Commit:
del main

Commit:
del main

Commit:
check

Commit:
check

Commit:
check

Commit:
check

Commit:
check

Commit:
check

Commit:
check

Commit:
check

Commit:
check

Commit:
check

Commit:
check

Commit:
check

Commit:
check me

hungry-honeybadger commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,133 @@
+	}
+	var r Data
+	for index, currentRune := range s {
+		if index == 0 && isDigit(currentRune) {
```
Эта проверка сработает всего 1 раз, а проверятся условие будет на каждой итерации поэтому лучше ее вынести за пределы цикла for.

hungry-honeybadger commented on hw02_unpack_string/unpack.go:
```diff
@@ -1,12 +1,133 @@
+}
+
+func parseValue(countRune interface{}) int {
+	switch v := countRune.(type) {
```
Эта задача не требует type switch. В данном случае type Switch ухудшает производительность и нарушает типизацию.

hungry-honeybadger reviewed (approved):
Хорошая работа. Отлично что для проверки на число используется не метод isDigit, а проверка диапазона руны. Это действительно ускоряет код. Оценка 8/10. Код очень сложно читать из-за того, что идет оперирование рунами как числами и в конце это приводит к тому, что нужен type switch. В итоге это не только бьет по читаемости но и по производительности.

Commit:
check me
